pkg/controllers/management/eks/eks_cluster_handler.go

```diff
@@ -282,7 +282,25 @@ func (e *eksOperatorController) onClusterChange(key string, cluster *mgmtv3.Clus
 			}
 		}
 
-		if cluster.Status.ServiceAccountToken == "" || cluster.Status.APIEndpoint == "" {
+		if cluster.Status.APIEndpoint == "" {
+			return e.recordCAAndAPIEndpoint(cluster)
+		}
+
+		if cluster.Status.EKSStatus.PrivateCanUsePublicAPI == nil && !*cluster.Status.EKSStatus.UpstreamSpec.PublicAccess {
+			// Check to see if we can still use the public API endpoint even though
+			// the cluster has private-only access
+			serviceToken, usePublicEndpoint := e.generateSATokenWithPublicAPI(cluster)
+			if usePublicEndpoint != nil {
+				cluster = cluster.DeepCopy()
+				cluster.Status.EKSStatus.PrivateCanUsePublicAPI = usePublicEndpoint
+				cluster.Status.ServiceAccountToken = serviceToken
+				return e.clusterClient.Update(cluster)
+			}
+			e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+			return cluster, nil
```

Comment:
It would be more conventional to just eval the error and if it's not dns then return the error since I believe that is the only time usePublicEndpoint would be nil. That way the error is logged somewhere and the controller will automatically retry. You could also just create an error here and return it if you do not think the error would contain anything useful. Manual enqueues should be avoided unless needed.